search/tasks_test.py

```diff
@@ -40,3 +56,180 @@ def test_increment_document_integer_field_task(mocked_api):
+    index_post_mock.si.assert_any_call(api_username=user.username, post_id=post1.id)
+    index_post_mock.si.assert_any_call(api_username=user.username, post_id=post2.id)
+    replace_mock.assert_called_once_with(group_mock.return_value)
+
+
+# pylint: disable=too-many-locals
+@pytest.mark.parametrize("temp_alias_exists", [True, False])
+def test_start_recreate_index(mocker, temp_alias_exists):
+    """
+    recreate_index should recreate the elasticsearch index and reindex all data with it
+    """
+    user = UserFactory.create()
+    client_mock = mocker.patch('channels.api.Api', autospec=True)
+    channel_names = ['a', 'b', 'c']
+    client_mock.return_value.list_channels.return_value = [
+        mocker.Mock(display_name=name) for name in channel_names
+    ]
+    mocker.patch('search.indexing_api.get_conn', autospec=True)
+    get_conn_mock = mocker.patch('search.tasks.get_conn', autospec=True)
+    index_channel_mock = mocker.patch('search.tasks.index_channel', autospec=True)
+    conn_mock = get_conn_mock.return_value
+    conn_mock.indices.exists_alias.return_value = temp_alias_exists
+    replace_mock = mocker.patch('celery.app.task.Task.replace', autospec=True, side_effect=TabError)
+    group_mock = mocker.patch('search.tasks.group', autospec=True)
+    chain_mock = mocker.patch('search.tasks.chain', autospec=True)
+    clear_and_create_mock = mocker.patch('search.tasks.clear_and_create_index', autospec=True)
+    finish_recreate_index_mock = mocker.patch('search.tasks.finish_recreate_index', autospec=True)
+
+    with pytest.raises(TabError):
+        start_recreate_index.delay(user.username)
+
+    reindexing_alias = get_reindexing_alias()
+    get_conn_mock.assert_called_once_with(verify=False)
+    assert clear_and_create_mock.call_count == 1
+    backing_index = clear_and_create_mock.call_args[1]['index_name']
+
+    conn_mock.indices.exists_alias.assert_called_once_with(name=reindexing_alias)
+    if temp_alias_exists:
+        conn_mock.indices.delete_alias.assert_any_call(
+            index="_all", name=reindexing_alias,
+        )
+    assert conn_mock.indices.delete_alias.called is temp_alias_exists
+
+    conn_mock.indices.put_alias.assert_called_once_with(
+        index=backing_index, name=reindexing_alias,
+    )
+    finish_recreate_index_mock.si.assert_called_once_with(backing_index)
+    assert group_mock.call_count == 1
+    list(group_mock.call_args[0][0])  # iterate through generator
+    for name in channel_names:
+        index_channel_mock.si.assert_any_call(api_username=user.username, channel_name=name)
+    chain_mock.assert_called_once_with(
+        group_mock.return_value,
+        finish_recreate_index_mock.si.return_value,
+    )
+    assert replace_mock.call_count == 1
+    assert replace_mock.call_args[0][1] == chain_mock.return_value
+
+
+@pytest.mark.parametrize("default_exists", [True, False])
+def test_finish_recreate_index(mocker, default_exists):
```

Comment:
This test case is really long, which makes the forked logic even tougher to sort out. You could potentially parametrize like this to get rid of at least some of the ifs:

```python
@pytest.mark.parametrize(
    "default_exists,backing_index_name,expected_alias_remove,expected_index_delete", [
        [True, 'old_index', True, True],
        [False, 'old_index', False, False],
    ]
)
```

Reply:
I'm not sure that parameterization would help. I moved the logic to test_switch_indices, can you take a second look?